File: java/billing-service/src/main/java/net/multi/language/billing/service/BillingService.java
package net.multi.language.billing.service;

import net.multi.language.billing.inbound.port.InboundMessagingPort;
import net.multi.language.billing.outbound.port.OutboundMessagingPort;
import net.multi.language.billing.inbound.port.InboundRestPort;
import net.multi.language.billing.outbound.port.OutboundRestPort;
import net.multi.language.common.dto.order.OrderChargingStatusDTO;
import net.multi.language.common.dto.order.PaymentDTO;
import net.multi.language.common.message.command.order.ChargeOrderCommand;
import net.multi.language.common.message.event.order.OrderChargedEvent;
import net.multi.language.common.message.event.order.OrderNotChargedEvent;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

@Service
public class BillingService implements InboundRestPort, InboundMessagingPort {

    private static final Logger LOGGER = LoggerFactory.getLogger(BillingService.class);
    private static final Random RANDOM = new Random(100);

    private final OutboundMessagingPort outboundMessagingPort;
    private final OutboundRestPort outboundRestPort;

    @Autowired
    public BillingService(final OutboundMessagingPort outboundMessagingPort, final OutboundRestPort outboundRestPort) {
        this.outboundMessagingPort = outboundMessagingPort;
        this.outboundRestPort = outboundRestPort;
    }

    @Override
    @Transactional
    public void chargeOrder(final ChargeOrderCommand chargeOrderCommand) {
        final long customerId = chargeOrderCommand.getCustomerId();
        final long orderId = chargeOrderCommand.getOrderId();
        final double orderTotal = chargeOrderCommand.getOrderTotal();

        LOGGER.info("Charging the customer with the ID {}, for the order with Id {}, for {} {}...", customerId, orderId,
                orderTotal, chargeOrderCommand.getCurrency());

        final int usedPaymentMethod = getPaymentMethod();
        final OrderChargingStatusDTO orderChargingStatus = outboundRestPort.charge(usedPaymentMethod, orderTotal);

        if (orderChargingStatus.isSuccessful()) {
            LOGGER.info("The customer {} was successfully charged for the order {}", customerId, orderId);
            outboundMessagingPort.publishOrderChargedEvent(createOrderChargedEvent(customerId, orderId));
        } else {
            final String failureReason = orderChargingStatus.getFailureReason()
                                                            .orElse("Cannot charge the card");
            LOGGER.warn("The customer {} could not be charged for the order {} - '{}'", customerId, orderId, failureReason);
            outboundMessagingPort.publishOrderNotChargedEvent(createOrderNotChargedEvent(customerId, orderId, failureReason));
        }
    }

    private OrderChargedEvent createOrderChargedEvent(long customerId, long orderId) {
        return new OrderChargedEvent(getNextMessageId(), getNextEventId(), customerId, orderId);
    }

    private OrderNotChargedEvent createOrderNotChargedEvent(long customerId, long orderId, String failureReason) {
        return new OrderNotChargedEvent(getNextMessageId(), getNextEventId(), customerId, orderId, failureReason);
    }

    @Override
    @Transactional
    public List<PaymentDTO> getPaymentsForCustomer(long customerId) {
        return new ArrayList<>();
    }

    private int getPaymentMethod() {
        // return the user's payment methods
        return RANDOM.nextInt(20);
    }

    private long getNextMessageId() {
        return RANDOM.nextLong();
    }

    private long getNextEventId() {
        // returned from the saved database event, before sending it (using transactional messaging)
        return RANDOM.nextLong();
    }
}

File: python/src/billing_service/services/billing_service.py
import logging
import os
from random import randint
from typing import List

from common.dto.order import PaymentDTO
from common.message.command.order import ChargeOrderCommand
from common.message.event.order import OrderChargedEvent, OrderNotChargedEvent

from billing_service.producer import Producer

from .payment_service import PaymentService

class BillingService:
    def __init__(
        self,
    ):
        self.logger = logging.getLogger(__name__)

    def charge_order(self, charge_order_command: ChargeOrderCommand):
        customer_id = charge_order_command.customer_id
        order_id = charge_order_command.order_id
        order_total = charge_order_command.order_total

        self.logger.info(
            f"Charging the customer with the ID {customer_id}, for the order with ID {order_id}, "
            f"for {order_total} {charge_order_command.currency}..."
        )

        used_payment_method = self.__get_payment_method()
        payment_service = PaymentService()
        order_charging_status = payment_service.charge(used_payment_method, order_total)
        producer = Producer(os.getenv("KAFKA_BROKERS", "localhost:9092"))

        if order_charging_status.successful:
            self.logger.info(
                f"The customer {customer_id} was successfully charged for the order {order_id}"
            )

            producer.publish_order_charged_event(
                OrderChargedEvent(
                    message_id=self.__get_next_message_id(),
                    event_id=self.__get_next_event_id(),
                    customer_id=customer_id,
                    order_id=order_id,
                )
            )
        else:
            failure_reason = (
                order_charging_status.failure_reason or "Cannot charge the card"
            )
            self.logger.warn(
                f"The customer {customer_id} could not be charged for the order {order_id} - "
                f"'{failure_reason}'"
            )

            producer.publish_order_not_charged_event(
                OrderNotChargedEvent(
                    message_id=self.__get_next_message_id(),
                    event_id=self.__get_next_event_id(),
                    customer_id=customer_id,
                    order_id=order_id,
                    reason=failure_reason,
                )
            )

    def charge_customer(self, customer_id: int, order_id: int) -> str:
        return (
            "The customer was charged"
            if randint(0, 1) == 0
            else "The charging has failed"
        )

    def get_payments_for_customer(self, customer_id: int) -> List[PaymentDTO]:
        return []

    def __get_payment_method(self) -> int:
        return randint(0, 19)

    def __get_next_message_id(self):
        # Generate a random message ID (for demonstration purposes)
        return randint(0, 2**64 - 1)

    def __get_next_event_id(self):
        # Generate a random event ID (for demonstration purposes)
        return randint(0, 2**64 - 1)
